fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/properties/FDBRecordContextProperties.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * FDBRecordContextProperties.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.properties;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class FDBRecordContextProperties {
+    private static final FDBRecordContextProperties EMPTY_PROPERTIES = new FDBRecordContextProperties(Collections.emptyMap());
+
+    @Nonnull
+    private final Map<String, FDBPropertyKey<?>> properties;
```

Comment:
I feel like I'm going back and forth on whether this really should be a `Map<String, PropertyKey>` or a `Map<PropertyKey, Object>` or even a `Map<PropertyKey, Supplier<Object>>`. The latter would let the user override a property value with something like:

```java
properties.put(propertyKey, newValue)
```

Or even:

```java
properties.put(propertyKey, () -> computePropertyValue());
```

Both of which could be could and I think (with the appropriate `put` methods) could even be made type safe. Then at property read time, the logic for `get` is something like "look in the map, and return the value in the map/compute the value from the supplier from the map if present; otherwise, use the value from the provided property key". But the benefit is the user doesn't need to go and create their own copy of the property key themselves.

However, I'm not really decided.

Reply:
With that implementation, I guess the supplier of ```FDBPropertyKey``` is likely not needed any more. Instead it just needs a default value. The user populates a map from ```FDBPropertyKey``` to ```Supplier<T>```. But in that case, the user has to be careful that the type argument of supplier needs to match with the key's type, and unfortunately there won't be a good way to check it before building the ```FDBRecordContextProperties``` as I can think of. Although the current implementation needs users to make sure the String key matches with the ```FDBPropertyKey```, but I guess it sounds like a little easier because users can directly do ```put(recordKey.getName, recordKey)``` and then they don't need to worry about type safe. 

I am open to further discussion on this topic.

Reply:
More detail here: https://github.com/FoundationDB/fdb-record-layer/pull/1477#discussion_r762137808

In my mind I was thinking of a `set`/`put` methods with signatures like:

```java
    Builder <T> set(FDBPropertyKey<T> property, T value);
    Builder <T> set(FDBPropertyKey<T> property, Supplier<T> value);
```

So, in this case, the thing that's guaranteeing type safety is that you can't provide a value (or supplier) that doesn't have the same type as the property key. The `get` method would likewise take an `FDBPropertyKey<T>` and would either return the value in the map (if set) or the default value (if not set). I think the `get` would need to do a runtime cast of the value (just like today), but I think the `put`-side validation would make it so that that runtime check always succeeds.